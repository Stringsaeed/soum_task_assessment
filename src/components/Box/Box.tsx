import { ReactElement, ReactNode, cloneElement, isValidElement } from 'react';
import { ColorValue, StyleSheet, ViewProps } from 'react-native';
import Animated, { AnimatedProps } from 'react-native-reanimated';

type Props = AnimatedProps<ViewProps> & {
  color?: ColorValue;
  foregroundColor?: ColorValue;
  size: number;
  children: ReactNode;
  factor?: number;
};

export default function Box({
  children,
  color,
  foregroundColor,
  size,
  factor = 0.5,
  style: overrideStyle,
  ...props
}: Props) {
  return (
    <Animated.View
      {...props}
      style={[
        styles.container,
        {
          backgroundColor: color,
          width: size,
        },
        overrideStyle,
      ]}>
      {isValidElement(children)
        ? cloneElement(children as ReactElement, {
            color: foregroundColor,
            size: Math.round(size * factor),
          })
        : children}
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  container: {
    aspectRatio: 1,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 8,
    alignSelf: 'flex-start',
    overflow: 'hidden',
  },
});
