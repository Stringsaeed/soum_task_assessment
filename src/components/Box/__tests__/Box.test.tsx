import { render, screen } from '@testing-library/react-native';
import { Check } from 'lucide-react-native';

import Box from '../Box';

describe('Box', () => {
  it('should render with default styles', () => {
    render(
      <Box testID="box" size={24}>
        <Check />
      </Box>,
    );

    const boxElement = screen.getByTestId('box');
    expect(boxElement).toBeDefined();
    expect(boxElement).toHaveStyle({ width: 24 });
  });

  it('should render with custom props', () => {
    render(
      <Box testID="box" color="blue" factor={0.8} size={24} foregroundColor="white">
        <Check testID="check-icon" />
      </Box>,
    );

    const boxElement = screen.getByTestId('box');
    expect(boxElement).toBeDefined();
    expect(boxElement).toHaveStyle({ backgroundColor: 'blue', width: 24 });

    const checkElement = screen.getAllByTestId('check-icon')[0];
    expect(screen.toJSON()).toMatchSnapshot();
    expect(checkElement).toBeDefined();
    expect(checkElement).toHaveProp('stroke', 'white');
    const iconSize = Math.round(24 * 0.8);
    expect(checkElement).toHaveStyle({ height: iconSize, width: iconSize });
  });
});
