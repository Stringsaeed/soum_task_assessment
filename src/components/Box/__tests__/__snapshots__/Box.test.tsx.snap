// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box should render with custom props 1`] = `
<View
  animatedStyle={
    {
      "value": {},
    }
  }
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "aspectRatio": 1,
      "backgroundColor": "blue",
      "borderRadius": 8,
      "justifyContent": "center",
      "overflow": "hidden",
      "width": 24,
    }
  }
  testID="box"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="19"
    bbWidth="19"
    fill="none"
    focusable={false}
    height={19}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="white"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 19,
          "width": 19,
        },
      ]
    }
    testID="check-icon"
    vbHeight={24}
    vbWidth={24}
    width={19}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 4294967295,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth="2"
    >
      <RNSVGPath
        d="M20 6 9 17l-5-5"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth="2"
        testID="check-icon"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
